data/migrations/2.4.1to2.4.2.sql

```diff
@@ -0,0 +1,21 @@
+--Ne plus modifier meta_date_create lors d'un UPDATE
+CREATE OR REPLACE FUNCTION public.fct_trg_meta_dates_change()
+  RETURNS trigger AS
+$BODY$
+begin
+        IF(TG_OP = 'UPDATE') THEN
```

Comment:
Seul le champs `meta_update_date` est rempli ici ?
Qu'en est-il du champs `meta_update_date`

Reply:
Ce qui est gênant dans le comportement actuel, c'est que la valeur des champs meta_date est forcé par le trigger, même si une valeur différente est explicitement indiqué lors des INSERT ou UPDATE.

En fait le but, c'est que le trigger ne soit pas appelé lors de l'insertion, car celui-ci modifie sournoisement meta_create_date si une valeur à été forcée lors de la requête INSERT.

Le champs **meta_create_date** est rempli avec `now()` (valeur par défaut) lors de la création (sans utilisation du trigger), sauf si valeur différent indiquée dans la requête. (champs pas affecté par le trigger donc)

Le champs **meta_update_date** est modifié à chaque modification de la ligne (par le trigger).


On pourra même aller plus loin et modifier le trigger pour ne PAS modifier meta_update_date si la valeur à été explicité lors de la requête UPDATE (par exemple si je veux faire des corrections minimes sans affecter la meta_update_date, ou forcé une meta_udate_date différente)

Reply:
Ok je vois. Il est possible que cette fonction soit utilisé par d'autre hors GeoNature. je fais le tour

Reply:
Ca marche. Si la valeur n'est pas forcée dans la requête, la PR ça ne devrait rien changer non ?

Reply:
Si pour ceux qui n'ont pas un `DEFAULT now()` sur le champ create

Reply:
```
        if(TG_OP = 'INSERT' AND NEW.meta_create_date IS NULL) THEN
                NEW.meta_create_date = NOW();
        ELSIF(TG_OP = 'UPDATE') THEN
                NEW.meta_update_date = NOW();
                if(NEW.meta_create_date IS NULL) THEN
                        NEW.meta_create_date = NOW();
                END IF;
        end IF;
        return NEW;
```
Est-ce que ça vous va ?

Reply:
J'ai peut-être raté un truc, mais j'ai vraiment du mal à comprendre l'intérêt d'un trigger pour la meta_create_date là ou `DEFAULT now()` sur _gn_synthese.synthese.meta_create_date_ fait parfaitement le job, et sans surprise. Là, le trigger va imposer une meta_create_date, même si la requête souhaite explicitement une date NULL (les champs meta_x_date étant pourtant NULLABLE)

Pour la meta_update_date, ok sur le principe, mais il y a une ligne en trop non ? (`NEW.meta_update_date = NOW();` avant le _if_ ). Je ne sais pas s'il y a moyen de checker si la valeur a été explicité ou non (je pense au cas où la requête UPDATE indiquerait explicitement `SET meta_update_date = NULL` , voir si pertinent ? )

Le fait de permettre un NULL pour meta_create_date ou meta_update_date peut être un autre sujet de débat, mais aujourd'hui la structure de la table le permet (et je trouve ça bien, pour justement retranscrire les cas où l'info est inconnu [imports vielle base par exemple])

Reply:
> J'ai peut-être raté un truc, mais j'ai vraiment du mal à comprendre l'intérêt d'un trigger pour la meta_create_date là ou DEFAULT now()

Parce qu'il y a d'autre tables qui utilise ce trigger et qui n'ont pas de `DEFAULT now()`. Donc le fait de modifier ce trigger implique une modification plus large d'autres tables. Qu'il faudrait faire... mais il faut le faire ! 

> et je trouve ça bien, pour justement retranscrire les cas où l'info est inconnu [imports vielle base par exemple]

Pour moi ces champs sont des champs de metadonnées GN. Ils indique le moment ou on a inséré ou modifié ces données dans GN, pas dans la base de données X ou Y dans un format différent. On ne devrait même pas pouvoir les éditer à mon sens.

Reply:
ok je comprend mieux, merci. Effectivement je viens de le voir tu t_datasets qui utilise aussi ce trigger. Je peux proposer une PR pour ajouter des `DEFAULT now()` sur les tables qui en manque, et ne plus appeler le trigger lors d'un insert sur ces tables.

Reply:
Et ca soulève un autre problème côté backend, la requête générée par l'ORM indique explicitement `meta_create_date=NULL`  lors de l'ajout d'un JDD, d'où l'utilité du trigger plutôt que du `DEFAULT now()`

Reply:
Je viens d'en discuter avec Amandine. Le plus propre serait de mettre ces champs en `NOT NULL default Now()`, et de prendre ta modif sur le trigger. Et oui il y a en plus le soucis de l'ORM à gérer.
Amandine a proposé de faire le tour quand elle a le temps.

Reply:
Dac', peut-être une piste ici pour l'ORM : https://docs.sqlalchemy.org/en/13/core/metadata.html#sqlalchemy.schema.Column.params.server_default